database/email.go

```diff
@@ -0,0 +1,128 @@
+func (db *DB) EmailCreate(ctx context.Context, m EmailMessage) error {
+	_, err := db.staticEmails.InsertOne(ctx, m)
+	if err != nil {
+		return errors.AddContext(err, "failed to Insert")
+	}
+	return nil
+}
+
+// FindEmails is a helper method that fetches emails and their ids from the
+// database.
+func (db *DB) FindEmails(ctx context.Context, filter bson.M, opts *options.FindOptions) ([]primitive.ObjectID, []EmailMessage, error) {
+	c, err := db.staticEmails.Find(ctx, filter, opts)
+	if err != nil {
+		return nil, nil, errors.AddContext(err, "failed to fetch ids")
+	}
+	defer func() {
+		if errDef := c.Close(ctx); errDef != nil {
+			err = errors.Compose(err, errors.AddContext(errDef, "error on closing DB cursor"))
+		}
+	}()
+	var ids []primitive.ObjectID
+	var msgs []EmailMessage
+	for c.Next(ctx) {
+		var m EmailMessage
+		if err = c.Decode(&m); err != nil {
+			return nil, nil, errors.AddContext(err, "failed to parse value from DB")
+		}
+		msgs = append(msgs, m)
+		ids = append(ids, m.ID)
+	}
+	return ids, msgs, nil
+}
+
+// MarkAsSent unlocks all given messages and marks them as sent.
+func (db *DB) MarkAsSent(ctx context.Context, ids []primitive.ObjectID) error {
+	if len(ids) == 0 {
+		return nil
+	}
+	filter := bson.M{"_id": bson.M{"$in": ids}}
+	update := bson.M{
+		"$set": bson.M{
+			"sent_at": time.Now().UTC(),
+		},
+	}
+	_, err := db.staticEmails.UpdateMany(ctx, filter, update)
+	if err != nil {
+		return errors.AddContext(err, "failed to mark emails as sent")
+	}
+	return nil
+}
+
+// MarkAsFailed increments the FailedAttemptsToSend counter on each message and
+// marks the message as Failed if that counter exceeds the maxAttemptsToSend.
+// It also unlocks all given messages.
+func (db *DB) MarkAsFailed(ctx context.Context, msgs []*EmailMessage) error {
+	if len(msgs) == 0 {
+		return nil
+	}
+	ids := make([]primitive.ObjectID, len(msgs))
+	var failed []primitive.ObjectID
+	var tryAgain []primitive.ObjectID
+	for i, m := range msgs {
+		ids[i] = m.ID
+		// the messages that are about to reach or exceed the limit will be
+		// marked as failed and won't be retries again.
+		if m.FailedAttemptsToSend >= maxAttemptsToSend-1 {
+			failed = append(failed, m.ID)
+		} else {
+			tryAgain = append(tryAgain, m.ID)
+		}
+	}
+
+	// Increment the counter on all messages that should be tried again.
+	filter := bson.M{"_id": bson.M{"$in": tryAgain}}
+	update := bson.M{
+		"$inc": bson.M{"failed_attempts_to_send": 1},
+	}
+	_, errInc := db.staticEmails.UpdateMany(ctx, filter, update)
+
+	// Mark all messages that exceeded the limit on failures as permanently
+	// failed.
+	var errFailed error
+	if len(failed) > 0 {
+		db.staticLogger.Warningf("%d email messages failed to be sent more than %d times and won't be tried anymore.", len(failed), maxAttemptsToSend)
+
+		filter = bson.M{"_id": bson.M{"$in": failed}}
+		update = bson.M{
+			"$inc": bson.M{"failed_attempts_to_send": 1},
```

Comment:
could've avoided having two fields by having `failed_attempts` and init it at zero, and then do `failed: {$lt: maxFailedAttempts}` in a predicate that fetches emails that need to be retried

Reply:
Good idea.